### PR TITLE
Cleanup Crucible resources with volume delete saga

### PR DESCRIPTION
**Pull request by jmpesp**
Add a volume delete saga, and call that whenever a volume needs to be deleted. Crucible resources now have accounting for their use, and when the number of references reaches zero then cleanup occurs.

Adds a table to store created region snapshots as part of this accounting, and make a record of when a Crucible snapshot is taken of a region. Region snapshots are uniquely identified by (dataset_id, region_id, snapshot_id), and additionally contain a snapshot address for identification as part of a volume construction request.

All DELETE calls to Crucible Agent(s) are now only performed by the volume delete saga.

Add tests for cases where if Nexus is properly accounting for snapshots. It should not delete the disk's corresponding region when the disk is deleted because a snapshot exists, or if multiple disks reference a running snapshot.

A small change to the simulated Crucible agent is also included in this commit - logic was moved out of the HTTP endpoint function into the Crucible struct. Logic doesn't belong in HTTP endpoint functions :)

Closes https://github.com/oxidecomputer/omicron/issues/1632

### CODE REVIEW

File: nexus/src/app/disk.rs [hunk 1]
```diff
@@ -510,6 +510,9 @@ impl super::Nexus {
             .project_delete_snapshot(opctx, &authz_snapshot, &db_snapshot)
             .await?;
 
+        // Kick off volume deletion saga
+        self.volume_delete(db_snapshot.volume_id).await?;
```
Comment (smklein) [inline nexus/src/app/disk.rs:514]: This is what I was mentioning earlier - I 100% understand that deleting a snapshot should release a reference to a volume, but other things may also reference that volume, no? Shouldn't this be only actually performing the volume deletion if the volume isn't referenced by anything else?

Comment (smklein) [reply nexus/src/app/disk.rs:514]: Okay, follow-up, I see I was being a dummy - this actually does exactly that, just in the saga.

We might want to mention - either via naming or a comment - that this isn't "unconditional volume deletion". It's really decrementing the volume refcount, which may or may not perform a deletion.

File: nexus/src/db/datastore/volume.rs [hunk 2]
```diff
@@ -86,4 +176,326 @@ impl DataStore {
...
+                        .get_result(conn)
+                        .optional()?;
+
+                    if volume.is_none() {
+                        // the volume was hard-deleted, return an empty
+                        // CrucibleResources
+                        return Ok(CrucibleResources::default());
+                    }
+
+                    let volume = volume.unwrap();
```
Comment (smklein) [inline nexus/src/db/datastore/volume.rs:285]: This should let us avoid `unwrap`:

```rust
let volume = if let Some(v) = volume {
  v
} else {
  return Ok(CrucibleResources::default());
}
```

**Review by smklein — APPROVED** [review]

File: nexus/src/app/sagas/volume_delete.rs [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    if request_count == 0 {
+        return Ok(());
+    }
+
+    futures::stream::iter(datasets_and_regions)
+        .map(|(dataset, region)| async move {
+            let url = format!("http://{}", dataset.address());
+            let client = CrucibleAgentClient::new(&url);
+            let id = RegionId(region.id().to_string());
+            client.region_delete(&id).await.map_err(|e| match e {
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/src/app/sagas/volume_delete.rs:343]: if you mean `region_delete`, yes it is.

File: nexus/src/app/sagas/volume_delete.rs [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+
+    // Send DELETE calls to the corresponding Crucible agents
+    delete_crucible_regions(
+        freed_datasets_regions_and_volumes
+            .iter()
+            .map(|(d, r, _)| (d.clone(), r.clone()))
+            .collect(),
+    )
+    .await
+    .map_err(ActionError::action_failed)?;
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/src/app/sagas/volume_delete.rs:269]: I believe so yes.

File: nexus/src/db/datastore/volume.rs (hunk 2 above)
Comment (jmpesp) [reply nexus/src/db/datastore/volume.rs]: done in 5979cd63

File: common/src/sql/dbinit.sql [hunk 5]
```diff
@@ -241,7 +280,18 @@ CREATE TABLE omicron.public.volume (
      * consumed by some Upstairs code to perform the volume creation. The Rust
      * type of this column should be Crucible::VolumeConstructionRequest.
      */
-    data TEXT NOT NULL
+    data TEXT NOT NULL,
+
+    /*
+     * A JSON document describing what resources to clean up when deleting this
+     * volume.
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply common/src/sql/dbinit.sql]: It didn't before 9692caca9a3578e6b3ceb0639babcb4819874a9e, but now it does :)

File: nexus/src/app/sagas/disk_delete.rs [hunk 6]
```diff
@@ -115,42 +85,13 @@ async fn sdd_delete_disk_record(
...
-}
-
 async fn sdd_delete_volume_record(
     sagactx: NexusActionContext,
 ) -> Result<(), ActionError> {
     let osagactx = sagactx.user_data();
     let volume_id = sagactx.lookup::<Uuid>("volume_id")?;
     osagactx
-        .datastore()
+        .nexus()
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/src/app/sagas/disk_delete.rs:94]: True, renamed in e2d7b1d2

File: nexus/src/db/datastore/region_snapshot.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+impl DataStore {
+    pub async fn region_snapshot_create(
+        &self,
+        region_snapshot: RegionSnapshot,
+    ) -> CreateResult<RegionSnapshot> {
+        use db::schema::region_snapshot::dsl;
+
+        // TODO https://github.com/oxidecomputer/omicron/issues/1168
+        diesel::insert_into(dsl::region_snapshot)
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/src/db/datastore/region_snapshot.rs]: I'd like to insert the RegionSnapshot, where if it exists already then do nothing, else insert it. I also wanted to return the RegionSnapshot, but it doesn't seem like this is possible. e300f8d2 has the correct behaviour and a test for it.

File: nexus/tests/integration_tests/volume_management.rs [hunk 8]
```diff
@@ -0,0 +1,1171 @@
...
+        let disk_url = format!("{}/{}", get_disks_url(), name);
+        NexusRequest::object_delete(client, &disk_url)
+            .authn_as(AuthnMode::PrivilegedUser)
+            .execute()
+            .await
+            .expect("failed to delete disk");
+    }
+
+    // Assert everything was cleaned up
+    assert!(disk_test.crucible_resources_deleted().await);
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/tests/integration_tests/volume_management.rs:1128]: I like this idea! done in 33468d4f21d817fe00dd98121590b78cb092fe7e

File: common/src/sql/dbinit.sql [hunk 9]
```diff
@@ -223,6 +223,45 @@ CREATE INDEX on omicron.public.Region (
...
+    dataset_id, region_id
+);
+
+/*
+ * Index on volume_references and snapshot_addr for crucible
+ * resource accounting lookup
+ */
+CREATE INDEX on omicron.public.region_snapshot (
+    volume_references
+);
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply common/src/sql/dbinit.sql:259]: No, we need to add to a `snapshot_addr`'s `volume_references`, whatever the value is

File: nexus/src/db/datastore/volume.rs [hunk 10]
```diff
@@ -8,59 +8,149 @@ use super::DataStore;
...
-           + diesel::query_builder::QueryId
-    where
-        T: Connection,
-    {
+    pub async fn volume_hard_delete(&self, volume_id: Uuid) -> DeleteResult {
         use db::schema::volume::dsl;
 
-        let now = Utc::now();
-        diesel::update(dsl::volume)
+        diesel::delete(dsl::volume)
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/src/db/datastore/volume.rs:152]: We're doing both - volumes are either soft deleted (if there are associated regions) and hard deleted (when those regions are cleaned up). I don't think I understand this comment though - `volume_hard_delete` doesn't filter on anything except the id, so it should find it every time.

File: nexus/src/app/disk.rs (hunk 1 above)
Comment (jmpesp) [reply nexus/src/app/disk.rs:514]: excellent point, expanded the comment in 62e54166

**Comment by jmpesp** [comment]
I believe I've addressed all the comments, let me know.

File: nexus/src/db/datastore/volume.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (smklein) [reply nexus/src/db/datastore/volume.rs:152]: I think my only beef here was with the error handling, where we handle the `NotFoundByLookup` case below. If we issue this delete request when the volume doesn't exist, will a `NotFound` error actually be returned?

https://docs.diesel.rs/master/diesel/result/enum.Error.html#variant.NotFound mentions this:

```
No rows were returned by a query expected to return at least one row.

This variant is only returned by get_result and first. load does not treat 0 rows as an error. If you would like to allow either 0 or 1 rows, call optional on the result.
```

I'm not sure what happens when we make this call with `execute` - it doesn't seem like an error for no rows to be updated.

**Review by smklein — APPROVED** [review]
Ship it! Thanks for the patches.

**Pull request merged**